x/reward/keeper/rules.go

```diff
@@ -0,0 +1,241 @@
+package keeper
+
+import (
+	"fmt"
+
+	authkeeper "github.com/cosmos/cosmos-sdk/x/auth/keeper"
+
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/provenance-io/provenance/x/reward/types"
+)
+
+func (k Keeper) ProcessTransactions(ctx sdk.Context) {
+	// Get all Active Reward Programs
+	rewardPrograms, err := k.GetAllActiveRewardPrograms(ctx)
+	if err != nil {
+		ctx.Logger().Error(err.Error())
+		return
+	}
+
+	// Grant shares for qualifying actions
+	for index := range rewardPrograms {
+		// Go through all the reward programs
+		actions, err := k.DetectQualifyingActions(ctx, &rewardPrograms[index])
+		if err != nil {
+			ctx.Logger().Error(err.Error())
+			continue
+		}
+
+		// Record any results
+		err = k.RewardShares(ctx, &rewardPrograms[index], actions)
+		if err != nil {
+			ctx.Logger().Error(err.Error())
+		}
+	}
+}
+
+// DetectQualifyingActions takes in the RewardProgram and checks if any of the qualifying actions is found in the event history
+func (k Keeper) DetectQualifyingActions(ctx sdk.Context, program *types.RewardProgram) ([]types.EvaluationResult, error) {
+	ctx.Logger().Info(fmt.Sprintf("EvaluateRules for RewardProgram: %d", program.GetId()))
+	results := []types.EvaluationResult(nil)
+
+	// Check if any of the transactions are qualifying actions
+	for _, supportedAction := range program.GetQualifyingActions() {
+		// Get the appropriate RewardAction
+		// If it's not supported we skip it
+		action, err := supportedAction.GetRewardAction(ctx)
+		if err != nil {
+			ctx.Logger().Info(err.Error())
+			continue
+		}
+
+		// Build all the qualifying actions from the abci events
+		actions, err := k.FindQualifyingActions(ctx, action)
+		if err != nil {
+			return nil, err
+		}
+
+		// Process actions and get the results
+		actions = k.ProcessQualifyingActions(ctx, program, action, actions)
+		results = append(results, actions...)
+	}
+
+	return results, nil
+}
+
+func (k Keeper) ProcessQualifyingActions(ctx sdk.Context, program *types.RewardProgram, processor types.RewardAction, actions []types.EvaluationResult) []types.EvaluationResult {
+	successfulActions := []types.EvaluationResult(nil)
+	if program == nil || processor == nil || actions == nil {
+		return successfulActions
+	}
+
+	for _, action := range actions {
+		state, err := k.GetRewardAccountState(ctx, program.GetId(), program.GetCurrentClaimPeriod(), action.Address.String())
+		if err != nil {
+			continue
+		}
+		if state.Validate() != nil {
+			state = types.NewRewardAccountState(program.GetId(), program.GetCurrentClaimPeriod(), action.Address.String(), 0, map[string]uint64{})
+		}
```

Comment:
We are creating a new state because of the pattern we have in place. When a key is not found we are returning a null/empty object. The id 0 is not considered valid so it fails validation, and we create the missing object. We are creating state on the fly because we won't know upfront how many people will be participating.

Reply:
In a nutshell it's valid, but we just haven't created the state yet for that user.